tangle-compiler/src/main/kotlin/tangle/inject/compiler/KotlinPoet.kt

```diff
@@ -111,19 +113,42 @@ fun List<AnnotationReference>.qualifierAnnotationSpecs(
   val fqName = annotationReference.fqName
 
   if (fqName == FqNames.inject) return@mapNotNull null
-
-  val qualifierAnnotation = fqName.toClassReference(module).annotations
-    .find { it.fqName == FqNames.qualifier }
-    ?: return@mapNotNull null
-
-  AnnotationSpec(qualifierAnnotation.classReference.asClassName()) {
-    qualifierAnnotation.arguments
+  if (
+    fqName.toClassReference(module)
+      .annotations
+      .none { it.fqName == FqNames.qualifier }
+  ) return@mapNotNull null
+
+  AnnotationSpec(annotationReference.classReference.asClassName()) {
+    annotationReference.arguments
       .forEach { arg ->
+        // Anvil's PSI Parsing for AnnotationArgumentReference.Psi currently does not fold
+        //   constants within string templates correctly, instead always taking only the first leaf
+        //   node. This causes an annotation like `@MyQualifier("Hello, $world")` to generate
+        //   `@MyQualifier("Hello, ")`, which is obviously incorrect. This could be resolved here
+        //   (in a separate PR) if desired, but it would be better to upstream the change.
+        //   See: https://github.com/square/anvil/blob/main/compiler-utils/src/main/java/com/squareup/anvil/compiler/internal/reference/AnnotationArgumentReference.kt#L267
+        if (
+          arg is AnnotationArgumentReference.Psi &&
+          arg.argument.stringTemplateExpression?.hasInterpolation() == true
+        ) {
+          throw TangleCompilationException(
+            buildString {
+              appendLine("String Interpolation in Qualifier Arguments is not currently supported")
+              appendLine("Here: ${arg.argument.text}")
+              annotationReference
+                .safeAs<AnnotationReference.Psi>()
+                ?.annotation?.text
+                ?.let { append("In: \"$it\"") }
+            },
+            element = arg.argument
+          )
+        }
+
         when (val value = arg.value<Any>()) {
           is KClassValue -> {
```

Comment:
Is there a reason that we are matchin on `KClassValue` and `EnumValue` here, it seems like [`arg.value<Any>()`](https://github.com/square/anvil/blob/main/compiler-utils/src/main/java/com/squareup/anvil/compiler/internal/reference/AnnotationArgumentReference.kt) won't ever return them, instead opting for giving us `ClassReference` and `FqName`.